lib/booking_manager.rb

```diff
@@ -0,0 +1,137 @@
+        check_in_date: input[:check_in_date],
+        check_out_date: input[:check_out_date]
+      }
+      reservation = Hotel::Reservation.new(reserve)
+      Hotel::Helper_Method.connect_reservation_to_room_and_sort(@rooms, input[:room_number], reservation)
+    end
+
+    def list_reservations(date)
+      reservations = []
+      @rooms.each do |room|
+        index = Hotel::Helper_Method.binary_search_reservations_return_index_if_found(room.reservations,date)
+        if index == nil
+          next
+        elsif
+          reservations << room.reservations[index]
+        end
+      end
+      if reservations.empty?
+        return raise StandardError, "No rooms reserved on provided date."
+      else
+        return reservations
+      end
+    end
+
+### BLOCK ROOM METHODS ###
+    def create_block(input)
+      check_in_date = input[:check_in_date]
+      check_out_date = input[:check_out_date]
+      vacant_rooms = []
+      vacant_rooms = search_all_rooms_availability(check_in_date, check_out_date)
+      number_of_rooms_to_block = input[:number_of_rooms_to_block]
+      if number_of_rooms_to_block > 5
+        return raise ArgumentError, 'Limit exceeded. Cannot create block larger than 5 rooms.'
+      end
+      blocked_rooms = []
+      number_of_rooms_to_block.times do |i|
+        block_room_reservation = {
+          room_number:  vacant_rooms[i],
+          check_in_date: check_in_date,
+          check_out_date: check_out_date,
+          block_name: input[:block_name],
+          block_discount: input[:block_discount]
+        }
+        reservation = Hotel::Block_Room.new(block_room_reservation)
+        Hotel::Helper_Method.connect_reservation_to_room_and_sort(@rooms, block_room_reservation[:room_number], reservation)
+
+        blocked_rooms << reservation
+      end
+      @block_reservations << blocked_rooms
+      Hotel::Helper_Method.sort_block_reservations_by_name(@block_reservations)
+    end
+
+    def reserve_room_in_block(input)
+      customer_name = input[:name]
+      available_rooms_in_block = check_block_status(input[:block_name])
+      available_rooms_in_block.first.name = customer_name
+      available_rooms_in_block.first.block_reservation_status = :BOOKED
+    end
+
+    def check_block_status(name_of_block)
+      available_rooms = []
+      name = name_of_block
+      block = @block_reservations.find { |block| block.first.block_name == name }
+      available_rooms = block.find_all { |reservation| reservation.block_reservation_status == :AVAILABLE }
+      if available_rooms.empty?
+        return raise StandardError, "All rooms in block booked."
+      else
+        return available_rooms
+      end
+    end
+
+### BOOKING UTILITY METHODS ###
+    def total_cost_of_stay(reservation)
+      cost = (reservation.nights_of_stay.length * ROOM_COST)
+      if reservation.class == Hotel::Block_Room
+        return (cost - (cost * reservation.block_discount))
+      else
+        return cost
+      end
```

Comment:
is there a way we can move the logic of calculating `cost` into the `reservation`? And if so, is there a way we can use inheritance to cover the case if reservation is a `Block_Room` too?